language-hask-source/src/Language/Hask/Parsed/AST.hs

```diff
@@ -0,0 +1,575 @@
+    newtypeConstructorName  :: ConstructorName
+  , newtypeConstructorLabel :: Maybe VariableName
+  , newtypeConstructorField :: Type
+  } deriving stock Show
+
+-- | Parsed "data" constructor.
+--
+-- > $dataConstructorName $dataConstructorFields
+data DataConstructor = DataConstructor{
+    dataConstructorName   :: ConstructorName
+  , dataConstructorFields :: Either [Type] [(VariableName, Strictness, Type)]
+    -- ^ "data" constructor fields, possibly named.
+    --
+    -- > $_ ...
+    --
+    -- or
+    --
+    -- > { $_ :: $_, ... }
+  } deriving stock Show
+
+-- | Parsed presence of strictness flag ("bang" @!@).
+type Strictness = Bool
+
+-- | Parsed class declaration.
+--
+-- > class $classType where $classDeclarations
+data Class = Class{
+    classType         :: Type
+    -- ^ See [Types in declaration heads](#TypesInDeclarationHeads).
+  , classDeclarations :: [Binding]
+  } deriving stock Show
+
+-- | Parsed instance declaration.
+--
+-- > instance $instanceType where $instanceDefinitions
+data Instance = Instance{
+    instanceType        :: Type
+    -- ^ See [Types in declaration heads](#TypesInDeclarationHeads).
+  , instanceDeclarations :: [Binding]
+    -- ^ See [Bindings in instances](#BindingsInInstances).
+  } deriving stock Show
+
+-- $BindingsInInstances
+-- == Bindings in instances #BindingsInInstances#
+-- We parse 'Binding's inside of instance declarations, which means that we can
+-- later fail with nicer error, plus we pave a way for possible future support
+-- for @InstanceSigs@ extension, which could provide value as a sort of
+-- "typechecked documentation".
+
+-- | Parsed bindings-related declaration.
+data Binding
+  = Signature [VariableName] Type
+    -- ^ Type signature for one or more bindings.
+    --
+    -- > $_, ... :: $2
+  | Fixity Associativity Natural [VariableName]
+    -- ^ Fixity declaration.
+    --
+    -- > $1 $2 $3
+  | ValueBindingDeclaration ValueBinding -- TODO: better name?
+  deriving stock Show
+
+-- | Parsed associativity in fixity declaration.
+--
+-- @infix@ or @infixl@ or @infixr@
+data Associativity = Infix | Infixl | Infixr deriving stock Show
+
+-- | Parsed value binding declaration.
+--
+-- > $bindingHead = $bindingBody where $bindingBindings
+data ValueBinding = ValueBinding{
+    bindingHead     :: Either Pattern (VariableName, NonEmpty Pattern)
+    -- ^ Head of either pattern or function binding.
+  , bindingBody     :: Expression
+  , bindingBindings :: Maybe [Binding]
+    -- ^ Optional (and possibly empty) @where@ clause.
+  } deriving stock Show
+
+-------------------------------------------------------------------------------
+-- TODO: how do we want to represent float literals?
+
+-- | Parsed expression. See 'Value'.
+type Expression = Value 'AsExpression
+
+-- | Parsed pattern. See 'Value'.
+type Pattern = Value 'AsPattern
+
+-- | Data kind restricting use of 'Value' constructors to expressions or
+-- patterns.
+data {-kind-} ValueUse = AsExpression | AsPattern
```

Comment:
Why does this have to be a data kind?

Reply:
You mean in comparison to
```hs
data AsExpression
data AsPattern
```
?

Reply:
Hm, I'm not so clear on what `ValueUse` represents, could you describe it?

Reply:
Sure - as mentioned above, `Expression` and `Pattern` are technically distinct types, but they share bunch of constructors (see those with return type `Value any`) - so instead of having `FooExpression` and `FooPattern` in distinct `data Expression` and `data Pattern`, we can create one shared `data Value` which contains both. Problem is, such type would be very imprecise when used directly in either of those contexts. So I've given it a type variable and constrained other constructors that appear only in one of the contexts with corresponding `ValueUse` - this way, you can write e.g.:
```hs
Lambda (Unit :| []) $ List [Integer 42] :: Expression
```
or
```hs
As (VariableName "x") $ Parens $ List [Successor (VariableName "n") 5] :: Pattern
```
You can't swap their types, because while
```hs
List [Integer 42]
```
and 
```hs
Unit
```
can be `Pattern`s, `Lambda` cannot. Similarly, `As` and `Successor` are restricted to `Pattern`, even though `Parens` and `List` aren't. This saves us a lot of wrapping/unwrapping with nested ADTs to get the similar effect.

Reply:
Thanks, that clears i up!